// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// The following only applies to changes made to this file as part of YugaByte development.
//
// Portions Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//
// Tests for the yb-admin command-line tool.

#include <boost/assign/list_of.hpp>
#include <gtest/gtest.h>

#include "yb/gutil/strings/split.h"
#include "yb/gutil/strings/substitute.h"

#include "yb/integration-tests/cluster_itest_util.h"
#include "yb/integration-tests/external_mini_cluster-itest-base.h"
#include "yb/integration-tests/test_workload.h"

#include "yb/tserver/tserver.pb.h"

#include "yb/util/path_util.h"
#include "yb/util/subprocess.h"

using boost::assign::list_of;
using yb::itest::TabletServerMap;
using yb::itest::TServerDetails;
using strings::Split;
using strings::Substitute;

constexpr int kTabletTimeout = 30;

namespace yb {
namespace tools {

static const char* const kTsCliToolName = "yb-ts-cli";

class YBTsCliTest : public ExternalMiniClusterITestBase {
 protected:
  // Figure out where the admin tool is.
  string GetTsCliToolPath() const;
};

string YBTsCliTest::GetTsCliToolPath() const {
  return GetToolPath(kTsCliToolName);
}

// Test deleting a tablet.
TEST_F(YBTsCliTest, TestDeleteTablet) {
  MonoDelta timeout = MonoDelta::FromSeconds(kTabletTimeout);
  std::vector<std::string> ts_flags = {
    "--enable_leader_failure_detection=false"s,
  };
  std::vector<std::string> master_flags = {
    "--catalog_manager_wait_for_new_tablets_to_elect_leader=false"s,
    "--use_create_table_leader_hint=false"s,
  };
  ASSERT_NO_FATALS(StartCluster(ts_flags, master_flags));

  TestWorkload workload(cluster_.get());
  workload.Setup(); // Easy way to create a new tablet.

  vector<tserver::ListTabletsResponsePB::StatusAndSchemaPB> tablets;
  for (const auto& entry : ts_map_) {
    TServerDetails* ts = entry.second.get();
    ASSERT_OK(itest::WaitForNumTabletsOnTS(ts, 1, timeout, &tablets));
  }
  string tablet_id = tablets[0].tablet_status().tablet_id();

  for (int i = 0; i < cluster_->num_tablet_servers(); i++) {
    ASSERT_OK(itest::WaitUntilTabletRunning(ts_map_[cluster_->tablet_server(i)->uuid()].get(),
                                            tablet_id, timeout));
  }

  string exe_path = GetTsCliToolPath();
  vector<string> argv;
  argv.push_back(exe_path);
  argv.push_back("--server_address");
  argv.push_back(yb::ToString(cluster_->tablet_server(0)->bound_rpc_addr()));
  argv.push_back("delete_tablet");
  argv.push_back(tablet_id);
  argv.push_back("Deleting for yb-ts-cli-test");
  ASSERT_OK(Subprocess::Call(argv));

  ASSERT_OK(inspect_->WaitForTabletDataStateOnTS(0, tablet_id, tablet::TABLET_DATA_TOMBSTONED));
  TServerDetails* ts = ts_map_[cluster_->tablet_server(0)->uuid()].get();
  ASSERT_OK(itest::WaitUntilTabletInState(ts, tablet_id, tablet::SHUTDOWN, timeout));
}

// Test readiness check before and after a tablet server is done bootstrapping.
TEST_F(YBTsCliTest, TestTabletServerReadiness) {
  MonoDelta timeout = MonoDelta::FromSeconds(kTabletTimeout);
  ASSERT_NO_FATALS(StartCluster({ "--TEST_tablet_bootstrap_delay_ms=2000"s }));

  TestWorkload workload(cluster_.get());
  workload.Setup(); // Easy way to create a new tablet.

  vector<tserver::ListTabletsResponsePB::StatusAndSchemaPB> tablets;
  for (const auto& entry : ts_map_) {
    TServerDetails* ts = entry.second.get();
    ASSERT_OK(itest::WaitForNumTabletsOnTS(ts, 1, timeout, &tablets));
  }
  string tablet_id = tablets[0].tablet_status().tablet_id();

  for (int i = 0; i < cluster_->num_tablet_servers(); i++) {
    ASSERT_OK(itest::WaitUntilTabletRunning(ts_map_[cluster_->tablet_server(i)->uuid()].get(),
                                            tablet_id, timeout));
  }

  ASSERT_NO_FATALS(cluster_->tablet_server(0)->Shutdown());
  ASSERT_OK(cluster_->tablet_server(0)->Restart());

  string exe_path = GetTsCliToolPath();
  vector<string> argv;
  argv.push_back(exe_path);
  argv.push_back("--server_address");
  argv.push_back(yb::ToString(cluster_->tablet_server(0)->bound_rpc_addr()));
  argv.push_back("is_server_ready");
  ASSERT_NOK(Subprocess::Call(argv));

  ASSERT_OK(WaitFor([&]() {
    return Subprocess::Call(argv).ok();
  }, MonoDelta::FromSeconds(10), "Wait for tablet bootstrap to finish"));
}

TEST_F(YBTsCliTest, TestRefreshFlags) {
  std::string gflag = "client_read_write_timeout_ms";
  std::string old_value = "120000";
  std::string new_value = "150000";

  // Write a flagfile.
  std::string flag_filename = JoinPathSegments(GetTestDataDirectory(), "flagfile.test");
  std::unique_ptr<WritableFile> flag_file;
  ASSERT_OK(Env::Default()->NewWritableFile(flag_filename, &flag_file));
  ASSERT_OK(flag_file->Append("--" + gflag + "=" + old_value));
  ASSERT_OK(flag_file->Close());

  // Start the cluster;
  vector<string> extra_flags = {"--flagfile=" + flag_filename};

  ASSERT_NO_FATALS(StartCluster(extra_flags, extra_flags));

  // Verify that the cluster is started with the custom GFlag value in the config.
  auto master_flag = ASSERT_RESULT(cluster_->master(0)->GetFlag(gflag));
  ASSERT_EQ(master_flag, old_value);
  auto ts_flag = ASSERT_RESULT(cluster_->tablet_server(0)->GetFlag(gflag));
  ASSERT_EQ(ts_flag, old_value);

  // Change the flagfile to have a different value for the GFlag.
  ASSERT_OK(Env::Default()->NewWritableFile(flag_filename, &flag_file));
  ASSERT_OK(flag_file->Append("--" + gflag + "=" + new_value));
  ASSERT_OK(flag_file->Close());

  // Send RefreshFlags RPC to the Master process
  string exe_path = GetTsCliToolPath();
  vector<string> argv;
  argv.push_back(exe_path);
  argv.push_back("--server_address");
  argv.push_back(yb::ToString(cluster_->master(0)->bound_rpc_addr()));
  argv.push_back("refresh_flags");
  ASSERT_OK(Subprocess::Call(argv));

  // Wait for the master process to have the updated GFlag value.
  ASSERT_OK(LoggedWaitFor([&]() -> Result<bool> {
    return VERIFY_RESULT(cluster_->master(0)->GetFlag(gflag)) == new_value;
  }, MonoDelta::FromSeconds(60), "Verify updated GFlag"));

  // The TServer should still have the old value because we didn't send it the RPC.
  ts_flag = ASSERT_RESULT(cluster_->tablet_server(0)->GetFlag(gflag));
  ASSERT_EQ(ts_flag, old_value);

  // Now, send a RefreshFlags RPC to the TServer process
  argv.clear();
  argv.push_back(exe_path);
  argv.push_back("--server_address");
  argv.push_back(yb::ToString(cluster_->tablet_server(0)->bound_rpc_addr()));
  argv.push_back("refresh_flags");
  ASSERT_OK(Subprocess::Call(argv));

  // Wait for the TS process to have the updated GFlag value.
  ASSERT_OK(LoggedWaitFor([&]() -> Result<bool> {
    return VERIFY_RESULT(cluster_->tablet_server(0)->GetFlag(gflag)) == new_value;
  }, MonoDelta::FromSeconds(60), "Verify updated GFlag"));
}

} // namespace tools
} // namespace yb
